### PR TITLE
Feat: get mlflow tracking uri and token in API

### PR DESCRIPTION
**Pull request by BenjaminMummery**
# The problem

We need a user-facing utility for constructing MLflow tracking URI and token, for all three execution environments, preferably with little-to-no defaults; should be a reliable basis for other QoL improvements. 

# This PR's solution

Implements 2 new functions: `sdk.mlflow.get_tracking_uri()` and `sdk.mlflow.get_tracking_token()`. Example usage:

```python
from orquestra import sdk
import os

os.environ["MLFLOW_TRACKING_URI"] = sdk.mlflow.get_tracking_uri(workspace_id="joka-123", config_name="prod-d", )
os.environ["MLFLOW_TRACKING_TOKEN"] = sdk.mlflow.get_tracking_token(config_name="prod-d")
```

## Behavior on Cluster/Studio

- The `config_name` parameter is ignored, with a warning to the user that this is the case.
- The token is read from the file specified by the `PASSPORT_FILE_ENV` environment variable
- The tracking URI is constructed by opening an HTTP session using the token in order to determine the namespace, combining this with the  `ORQ_MLFLOW_CR_NAME` and `ORQ_MLFLOW_PORT` environment variables in the form `http://[ORQ_MLFLOW_CR_NAME].[NAMESPACE]:[ORQ_MLFLOW_PORT]`

## Behaviour on Local

- The token is sourced from the stored config.
- The tracking URI is constructed from the stored config URI and workspace ID in the form `[URI]/mlflow/[WORKSPACE ID]`

## A Note on Determining the Execution Environment

At present we use two mechanisms to determine whether the execution environment is 'local' or 'remote'. 
If the `CURRENT_CLUSTER_ENV` environment variable is set, we take this as concrete proof that we're remote. As this variable was only added in [OP-137](https://zapatacomputing.atlassian.net/browse/OP-137), its absence cannot be assumed to be proof that we're running locally. Therefore we also check the `PASSPORT_FILE_ENV`, `MLFLOW_CR_NAME`, `MLFLOW_PORT`, and `MLFLOW_ARTIFACTS_DIR` environment variables. If one or more of these are not set, we take this as indicating that we're running on local. This is a temporary measure as we only support N-1 backwards compatibility, so this second layer of checking can be removed in the next release ([ORQSDK-914](https://zapatacomputing.atlassian.net/browse/ORQSDK-914))

# Checklist

_Check that this PR satisfies the following items:_

- [x] Tests have been added for new features/changed behavior (if no new features have been added, check the box).
- [x] The [changelog file](CHANGELOG.md) has been updated with a user-readable description of the changes (if the change isn't visible to the user in any way, check the box).
- [X] The PR's title is prefixed with `<feat/fix/chore/imp[rovement]/int[ernal]/docs>[!]:`
- [X] The PR is linked to a JIRA ticket (if there's no suitable ticket, check the box).

[ORQSDK-879]

[ORQSDK-879]: https://zapatacomputing.atlassian.net/browse/ORQSDK-879?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[OP-137]: https://zapatacomputing.atlassian.net/browse/OP-137?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/orquestra/sdk/mlflow/_connection_utils.py [hunk 1]
```diff
@@ -5,11 +5,128 @@
...
 DEFAULT_TEMP_ARTIFACTS_DIR: Path = ORQUESTRA_BASE_PATH / "mlflow" / "artifacts"
+RESOURCE_CATALOG_URI: str = "http://orquestra-resource-catalog.resource-catalog"
+
+
+# region: private
+def _is_executing_remoteley() -> bool:
+    """
+    Determine whether the code is being executed locally, or on a cluster/studio.
+    """
+    # TODO: at present this uses the platform-specific environment variables, assuming
```
Comment (alexjuda) [inline src/orquestra/sdk/mlflow/_connection_utils.py:28]: **nitpick**: please link an SDK JIRA ticket so we don't have a hanging TODO

File: src/orquestra/sdk/mlflow/_connection_utils.py [hunk 2]
```diff
@@ -22,15 +139,95 @@ def get_temp_artifacts_dir() -> Path:
...
+    # locally and use the default artifacts dir.
+
+    temp_dir_path.mkdir(parents=True, exist_ok=True)
+
+    return temp_dir_path
+
+
+def get_tracking_uri(workspace_id: str, config_name: Optional[str] = None) -> str:
+    """
+    Get the MLFlow tracking URI for the specified workspace.
```
Comment (alexjuda) [inline src/orquestra/sdk/mlflow/_connection_utils.py:156]: **nitpick**: if it exists, a docstring should convey more information than what can be read from the signature alone. WDYT?
```suggestion
    Infer a URI for accessing an MLflow tracking server deployed within this workspace.
    
    When run within an Orquestra cluster, this function returns an "internal URI" that helps save cluster bandwidth. This works even without specifying ``config_name``.
    
    When run outside a cluster, this function returns an "external URI". Requires passing ``config_name``.
```

File: src/orquestra/sdk/mlflow/_connection_utils.py [hunk 3]
```diff
@@ -22,15 +139,95 @@ def get_temp_artifacts_dir() -> Path:
...
+
+    Args:
+        workspace_id: ID of the workspace.
```
Comment (alexjuda) [inline src/orquestra/sdk/mlflow/_connection_utils.py:159]: **suggestion**: 
```suggestion
        workspace_id: ID of the workspace. You can get it from Orquestra Portal UI.
```

File: src/orquestra/sdk/mlflow/_connection_utils.py [hunk 4]
```diff
@@ -22,15 +139,95 @@ def get_temp_artifacts_dir() -> Path:
...
+                f"The config '{config_name}' has no URI associated with it. "
+                "In order to determine the tracking URI, the config must specify a "
+                "remote execution environment where the MLFlow instance is hosted. "
+                "Please double-check the config name and try again."
+            ) from e
+
+
+def get_tracking_token(config_name: Optional[str] = None) -> str:
+    """
+    Get the MLFlow tracking token.
```
Comment (alexjuda) [inline src/orquestra/sdk/mlflow/_connection_utils.py:205]: **nitpick**: ditto docstring-vs-signature

```suggestion
    Get a token suitable for authorizing the MLflow client for accessing Orquestra's MLflow tracking server.
```

File: tests/sdk/v2/mlflow/test_connection_utils.py [hunk 5]
```diff
@@ -65,3 +76,386 @@ def test_creates_dir(tmp_path: pathlib.Path, monkeypatch):
...
+
+class TestGetMLFlowCRNameAndPort:
+    @staticmethod
+    def test_happy_path(monkeypatch: MonkeyPatch, mock_is_executing_remotely):
+        # Given
+        monkeypatch.setenv(MLFLOW_CR_NAME, "<mlflow cr name sentinel>")
+        monkeypatch.setenv(MLFLOW_PORT, "<mlflow port sentinel>")
+
+        # When
+        name, port = sdk.mlflow._connection_utils._get_mlflow_cr_name_and_port()
```
Comment (alexjuda) [inline tests/sdk/v2/mlflow/test_connection_utils.py:105]: **nitpick**: it's a good practice not to test private functions
```suggestion
        name, port = sdk.mlflow._connection_utils.get_mlflow_cr_name_and_port()
```

File: tests/sdk/v2/mlflow/test_connection_utils.py [hunk 6]
```diff
@@ -65,3 +76,386 @@ def test_creates_dir(tmp_path: pathlib.Path, monkeypatch):
...
+
+        # Then
+        assert token == "<token sentinel>"
+
+
+class TestMakeWorkspaceZRI:
+    @staticmethod
+    def test_string_construction():
+        # When
+        zri = sdk.mlflow._connection_utils._make_workspace_zri(
```
Comment (alexjuda) [inline tests/sdk/v2/mlflow/test_connection_utils.py:146]: **nitpick**: ditto testing private functions

File: src/orquestra/sdk/mlflow/_connection_utils.py [hunk 7]
```diff
@@ -5,11 +5,128 @@
...
+    # that time.
+    envvars = [
+        _env.PASSPORT_FILE_ENV,
+        _env.MLFLOW_CR_NAME,
+        _env.MLFLOW_PORT,
+        _env.MLFLOW_ARTIFACTS_DIR,
+    ]
+    if None in [os.getenv(envvar) for envvar in envvars]:
+        return False
+    return True
```
Comment (alexjuda) [inline src/orquestra/sdk/mlflow/_connection_utils.py:43]: **nitpick**: missing unit tests

File: src/orquestra/sdk/mlflow/_connection_utils.py [hunk 8]
```diff
@@ -5,11 +5,128 @@
...
+    """
+    Reads in the token.
+
+    The file path is specified by the PASSPOT_FILE_ENV environment variable.
+
+    Raises:
+        EnvironmentError: when the PASSPORT_FILE_ENV environment variable is not set.
+    """
+    if not (passport_file_path := os.getenv(_env.PASSPORT_FILE_ENV)):
+        raise EnvironmentError(
```
Comment (alexjuda) [inline src/orquestra/sdk/mlflow/_connection_utils.py:90]: **nitpick**: missing unit tests

**Review by alexjuda — APPROVED** [review]
**praise**: that's a very helpful PR description! I esp like the consideration how this affects execution in different environments.

The code mostly LGTM. Added a bunch of non-critical nitpicks

File: src/orquestra/sdk/mlflow/_connection_utils.py (hunk 1 above)
Comment (SebastianMorawiec) [reply src/orquestra/sdk/mlflow/_connection_utils.py:28]: I think according to this post:https://zapatista.slack.com/archives/C02DN7BA04E/p1689085958796819 we already have ORQ_IN_CLUSTER, right?

File: src/orquestra/sdk/mlflow/_connection_utils.py [hunk 9]
```diff
@@ -5,11 +5,128 @@
...
+            f"The '{_env.PASSPORT_FILE_ENV}' environment variable is not set."
+        )
+    return Path(passport_file_path).read_text()
+
+
+def _make_workspace_zri(workspace_id: str) -> str:
```
Comment (SebastianMorawiec) [inline src/orquestra/sdk/mlflow/_connection_utils.py:96]: Suggestion: Im pretty sure we have this logic somewhere in the code already. Could we reuse it?

Comment (BenjaminMummery) [reply src/orquestra/sdk/mlflow/_connection_utils.py]: Good idea, I've moved it to the spaces api.

File: src/orquestra/sdk/mlflow/_connection_utils.py [hunk 10]
```diff
@@ -6,15 +6,111 @@
...
+        # TODO: https://zapatacomputing.atlassian.net/browse/ORQSDK-914 - this
+        # workaround exists to maintian compatibility with previous WDR versions.
+        # Future releases should remove this else block.
+        envvars = [
+            _env.PASSPORT_FILE_ENV,
+            _env.MLFLOW_CR_NAME,
+            _env.MLFLOW_PORT,
+            _env.MLFLOW_ARTIFACTS_DIR,
+        ]
+        if None not in [os.getenv(envvar) for envvar in envvars]:
```
Comment (alexjuda) [inline src/orquestra/sdk/mlflow/_connection_utils.py:46]: **nitpick**: when reading this line I find it difficult to figure out the author's intention, there's just too many negations! WDYT about the following, does it capture the same thing?
```suggestion
        if all(os.getenv(envvar) for envvar in envvars):
```

**Review by alexjuda — APPROVED** [review]

**Pull request merged**